torch_xla/experimental/xla_sharded_tensor.py

```diff
@@ -0,0 +1,106 @@
+import torch
+from torch.utils._pytree import tree_map
+import torch_xla
+
+from dataclasses import dataclass
+from typing import List, Tuple, Iterator
+import contextlib
+import collections
+
+
+@dataclass
+class XLAShard:
+  data: torch.Tensor
+  rank: int
+
+
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+  guard = torch._C._DisableTorchDispatch()  # type: ignore[attr-defined]
+  try:
+    yield
+  finally:
+    del guard
+
+
+class XLAShardedTensor(torch.Tensor):
```

Comment:
Looking at the code `XLAShardedTensor` wraps around a `torch.Tensor` but it doesn't seems to add any functionality besides having a `sharding_spec`. `mark_sharding` is also called separately from the construction of the `XLAShardedTensor`. Is the purpose of this class mostly for dispatch? However I didn't see `__torch_dispatch__` being implemented either.

Reply:
So the key and inherent property of `XLAShardedTensor` is that it behaves just like `torch.Tensor` and dispatched (`__torch_dispatch__` is working) to the `xla` backend as `XlaTensor` with sharding annotations; so, we don't have to build a separate set of aten ops, xla lowering ops for it. While wrapping and appearing as the input `torch.Tensor`, `XLAShardedTensor` will provide a list of shards as individual tensors -- but this part is not implemented, yet.